service/legacy_ui_redirects.go

```diff
@@ -0,0 +1,77 @@
+package service
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/evergreen-ci/gimlet"
+)
+
+func (uis *UIServer) legacyWaterfallPage(w http.ResponseWriter, r *http.Request) {
+	projCtx := MustHaveProjectContext(r)
+	project, err := projCtx.GetProject(r.Context())
+
+	if err != nil || project == nil {
+		http.Redirect(w, r, uis.Settings.Ui.UIv2Url, http.StatusMovedPermanently)
+		return
+	}
+
+	newUIURL := fmt.Sprintf("%s/project/%s/waterfall", uis.Settings.Ui.UIv2Url, project.Identifier)
+	http.Redirect(w, r, newUIURL, http.StatusMovedPermanently)
+}
+
+func (uis *UIServer) legacyDistrosPage(w http.ResponseWriter, r *http.Request) {
+	spruceLink := fmt.Sprintf("%s/distros", uis.Settings.Ui.UIv2Url)
+	http.Redirect(w, r, spruceLink, http.StatusPermanentRedirect)
+}
+
+func (uis *UIServer) legacyTaskQueue(w http.ResponseWriter, r *http.Request) {
+	distro := gimlet.GetVars(r)["distro"]
+	taskId := gimlet.GetVars(r)["task_id"]
+	newUILink := ""
+	if len(uis.Settings.Ui.UIv2Url) > 0 {
+		newUILink = fmt.Sprintf("%s/task-queue/%s/%s", uis.Settings.Ui.UIv2Url, distro, taskId)
+	}
+
+	http.Redirect(w, r, newUILink, http.StatusTemporaryRedirect)
+}
+
+func (uis *UIServer) legacyTaskHistoryPage(w http.ResponseWriter, r *http.Request) {
+	projCtx := MustHaveProjectContext(r)
+	executionStr := gimlet.GetVars(r)["execution"]
+	var execution int
+	var err error
+
+	if executionStr != "" {
+		execution, err = strconv.Atoi(executionStr)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Bad execution number: %v", executionStr), http.StatusBadRequest)
+			return
+		}
+	}
+	if projCtx.Task == nil {
+		http.Error(w, "Not found", http.StatusNotFound)
+		return
+	}
+	http.Redirect(w, r, fmt.Sprintf("%s/task/%s/history?execution=%d", uis.Settings.Ui.UIv2Url, projCtx.Task.Id, execution), http.StatusPermanentRedirect)
+}
+
+func (uis *UIServer) legacyTaskPage(w http.ResponseWriter, r *http.Request) {
+	projCtx := MustHaveProjectContext(r)
+	project, err := projCtx.GetProject(r.Context())
+
+	if err != nil || project == nil {
+		http.Error(w, "not found", http.StatusNotFound)
+		return
+	}
+	// There's no longer an equivalent Task History page on Spruce, so we have to link to a different page. Waterfall is used since
+	// it is the most relevant.
+	http.Redirect(w, r, fmt.Sprintf("%s/project/%s/waterfall", uis.Settings.Ui.UIv2Url, project.Identifier), http.StatusPermanentRedirect)
```

Comment:
I see the bug here. This is redirecting to the waterfall when it should be redirecting to the spruce task page.